### ISSUE TITLE
YCSB - REDIS ERROR

### ISSUE THREAD

**Issue by vishakha94** [brianfrankcooper/YCSB#619]
Exception in thread "Thread-1" java.lang.IllegalArgumentException: hostname can't be null

Hey, I am getting this error when I am trying to load the data using this command: "./bin/ycsb load redis -s -P workloads/workloada > outputLoad.txt"

Any pointer?

Thank you!

**Comment by vishakha94** [comment]
I also tried replacing redis.host=127.0.0.1 to localhost.

**Comment by cmccoy** [comment]
Where are you defining redis.host?

**Comment by vishakha94** [comment]
I used this command: "./bin/ycsb load redis -s -P workloads/workloada -p "redis.host=127.0.0.1" -p "redis.port=6379" > outputLoad.txt"

**Comment by cmccoy** [comment]
Is there anything else in your traceback?

**Comment by vishakha94** [comment]
Starting test.
2016-02-09 19:33:37:516 0 sec: 0 operations; est completion in 0 seconds 
DBWrapper: report latency for each error is false and specific error codes to track for latency are: []
2016-02-09 19:33:37:986 0 sec: 1000 operations; 1988.07 current ops/sec; [CLEANUP: Count=1, Max=134, Min=134, Avg=134, 90=134, 99=134, 99.9=134, 99.99=134] [INSERT: Count=1000, Max=6907, Min=233, Avg=365.12, 90=435, 99=1119, 99.9=4575, 99.99=6907] 
  This is what I get when I run the command to set configs

**Comment by vishakha94** [comment]
Loading workload...
Starting test.
Exception in thread "Thread-1" java.lang.IllegalArgumentException: hostname can't be null
	at java.net.InetSocketAddress.checkHost(InetSocketAddress.java:149)
	at java.net.InetSocketAddress.<init>(InetSocketAddress.java:216)
	at redis.clients.jedis.Connection.connect(Connection.java:119)
	at redis.clients.jedis.BinaryClient.connect(BinaryClient.java:54)
	at redis.clients.jedis.BinaryJedis.connect(BinaryJedis.java:1657)
	at com.yahoo.ycsb.db.RedisClient.init(RedisClient.java:71)
	at com.yahoo.ycsb.DBWrapper.init(DBWrapper.java:77)
	at com.yahoo.ycsb.ClientThread.run(Client.java:289)
2016-02-09 19:37:05:617 0 sec: 0 operations; est completion in 0 seconds 
2016-02-09 19:37:05:620 0 sec: 0 operations; est completion in 106751991167300 days 15 hours

**Comment by kruthar** [comment]
@vishakha94 - It looks to me like you are mixing up the commands and the output you are posting here.

>Exception in thread "Thread-1" java.lang.IllegalArgumentException: hostname can't be null

This error gets thrown if you *DO NOT* specify ```redis.host```.

This is the error you should get if you specify ```redis.host``` and it is incorrect:

```
Exception in thread "Thread-1" redis.clients.jedis.exceptions.JedisConnectionException: java.net.ConnectException: Connection refused
```

This is not the error you posted, and so does not seem to match if you actually ran the command below and the host was incorrect:

>I used this command: "./bin/ycsb load redis -s -P workloads/workloada -p "redis.host=127.0.0.1" -p "redis.port=6379" > outputLoad.txt"

This output you posted:

>Starting test.
2016-02-09 19:33:37:516 0 sec: 0 operations; est completion in 0 seconds 
DBWrapper: report latency for each error is false and specific error codes to track for latency are: []
2016-02-09 19:33:37:986 0 sec: 1000 operations; 1988.07 current ops/sec; [CLEANUP: Count=1, Max=134, Min=134, Avg=134, 90=134, 99=134, 99.9=134, 99.99=134] [INSERT: Count=1000, Max=6907, Min=233, Avg=365.12, 90=435, 99=1119, 99.9=4575, 99.99=6907] 

...looks correct. It says that there were 1000 Insert operations which matches the default ```load``` configurations for workloada.

In short, I think you actually solved your problem by specifiying ```redis.host``` and ```redis.port```. But if you are still having issues please post your entire command session starting with the YCSB command you run and followed by the ENTIRE ycsb output. *DO NOT* redirect output to a file so that all of your output goes to stdout and you can post it here.

**Comment by vishakha94** [comment]
Ok. So in my subsequent comments, I am gonna post the outputs of all my commands run sequentially.

**Comment by vishakha94** [comment]
**mvn -pl com.yahoo.ycsb:redis-binding -am clean package**
The output for this command is way too long. So I shall just post its last part.
Output:
-------------------------------------------------------
 T E S T S
-------------------------------------------------------

Results :

Tests run: 0, Failures: 0, Errors: 0, Skipped: 0

[INFO] 
[INFO] --- maven-jar-plugin:2.2:jar (default-jar) @ redis-binding ---
[INFO] Building jar: /home/visheei/YCSB/redis/target/redis-binding-0.7.0-SNAPSHOT.jar
[INFO] 
[INFO] --- maven-dependency-plugin:2.10:copy-dependencies (stage-dependencies) @ redis-binding ---
[INFO] Copying commons-pool-1.5.5.jar to /home/visheei/YCSB/redis/target/dependency/commons-pool-1.5.5.jar
[INFO] Copying jedis-2.0.0.jar to /home/visheei/YCSB/redis/target/dependency/jedis-2.0.0.jar
[INFO] 
[INFO] --- maven-assembly-plugin:2.5.5:single (default) @ redis-binding ---
[WARNING] The following patterns were never triggered in this artifact exclusion filter:
o  '*:sources'

[INFO] Building tar: /home/visheei/YCSB/redis/target/ycsb-redis-binding-0.7.0-SNAPSHOT.tar.gz
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO] 
[INFO] YCSB Root ......................................... SUCCESS [2.316s]
[INFO] Core YCSB ......................................... SUCCESS [8.265s]
[INFO] Per Datastore Binding descriptor .................. SUCCESS [0.468s]
[INFO] YCSB Datastore Binding Parent ..................... SUCCESS [0.642s]
[INFO] Redis DB Binding .................................. SUCCESS [1.253s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 13.567s
[INFO] Finished at: Tue Feb 09 21:20:54 EST 2016
[INFO] Final Memory: 31M/416M
[INFO] ------------------------------------------------------------------------

**Comment by vishakha94** [comment]
**./bin/ycsb load redis -s -P workloads/workloada -p "redis.host=127.0.0.1" -p "redis.port=6379" 
**
OUTPUT:
[WARN]  Running against a source checkout. In order to get our runtime dependencies we'll have to invoke Maven. Depending on the state of your system, this may take ~30-45 seconds
[DEBUG]  Running 'mvn -pl com.yahoo.ycsb:redis-binding -am package -DskipTests dependency:build-classpath -DincludeScope=compile -Dmdep.outputFilterFile=true'
java -cp /home/visheei/YCSB/redis/conf:/home/visheei/YCSB/redis/target/redis-binding-0.7.0-SNAPSHOT.jar:/home/visheei/.m2/repository/commons-pool/commons-pool/1.5.5/commons-pool-1.5.5.jar:/home/visheei/YCSB/core/target/core-0.7.0-SNAPSHOT.jar:/home/visheei/.m2/repository/redis/clients/jedis/2.0.0/jedis-2.0.0.jar:/home/visheei/.m2/repository/org/codehaus/jackson/jackson-core-asl/1.9.4/jackson-core-asl-1.9.4.jar:/home/visheei/.m2/repository/org/hdrhistogram/HdrHistogram/2.1.4/HdrHistogram-2.1.4.jar:/home/visheei/.m2/repository/org/codehaus/jackson/jackson-mapper-asl/1.9.4/jackson-mapper-asl-1.9.4.jar com.yahoo.ycsb.Client -db com.yahoo.ycsb.db.RedisClient -s -P workloads/workloada -p redis.host=127.0.0.1 -p redis.port=6379 -load
YCSB Client 0.1
Command line: -db com.yahoo.ycsb.db.RedisClient -s -P workloads/workloada -p redis.host=127.0.0.1 -p redis.port=6379 -load
Loading workload...
Starting test.
DBWrapper: report latency for each error is false and specific error codes to track for latency are: []
2016-02-09 21:25:48:393 0 sec: 0 operations; est completion in 0 seconds 
2016-02-09 21:25:49:236 0 sec: 1000 operations; 1098.9 current ops/sec; [CLEANUP: Count=1, Max=217, Min=217, Avg=217, 90=217, 99=217, 99.9=217, 99.99=217] [INSERT: Count=1000, Max=10927, Min=431, Avg=710.71, 90=887, 99=1988, 99.9=7059, 99.99=10927] 
[OVERALL], RunTime(ms), 910.0
[OVERALL], Throughput(ops/sec), 1098.901098901099
[CLEANUP], Operations, 1.0
[CLEANUP], AverageLatency(us), 217.0
[CLEANUP], MinLatency(us), 217.0
[CLEANUP], MaxLatency(us), 217.0
[CLEANUP], 95thPercentileLatency(us), 217.0
[CLEANUP], 99thPercentileLatency(us), 217.0
[INSERT], Operations, 1000.0
[INSERT], AverageLatency(us), 710.712
[INSERT], MinLatency(us), 431.0
[INSERT], MaxLatency(us), 10927.0
[INSERT], 95thPercentileLatency(us), 1506.0
[INSERT], 99thPercentileLatency(us), 1988.0
[INSERT], Return=OK, 1000

**Comment by vishakha94** [comment]
**./bin/ycsb load redis -s -P workloads/workloada **

[WARN]  Running against a source checkout. In order to get our runtime dependencies we'll have to invoke Maven. Depending on the state of your system, this may take ~30-45 seconds
[DEBUG]  Running 'mvn -pl com.yahoo.ycsb:redis-binding -am package -DskipTests dependency:build-classpath -DincludeScope=compile -Dmdep.outputFilterFile=true'
java -cp /home/visheei/YCSB/redis/conf:/home/visheei/YCSB/redis/target/redis-binding-0.7.0-SNAPSHOT.jar:/home/visheei/.m2/repository/commons-pool/commons-pool/1.5.5/commons-pool-1.5.5.jar:/home/visheei/YCSB/core/target/core-0.7.0-SNAPSHOT.jar:/home/visheei/.m2/repository/redis/clients/jedis/2.0.0/jedis-2.0.0.jar:/home/visheei/.m2/repository/org/codehaus/jackson/jackson-core-asl/1.9.4/jackson-core-asl-1.9.4.jar:/home/visheei/.m2/repository/org/hdrhistogram/HdrHistogram/2.1.4/HdrHistogram-2.1.4.jar:/home/visheei/.m2/repository/org/codehaus/jackson/jackson-mapper-asl/1.9.4/jackson-mapper-asl-1.9.4.jar com.yahoo.ycsb.Client -db com.yahoo.ycsb.db.RedisClient -s -P workloads/workloada -load
YCSB Client 0.1
Command line: -db com.yahoo.ycsb.db.RedisClient -s -P workloads/workloada -load
Loading workload...
Starting test.
Exception in thread "Thread-1" java.lang.IllegalArgumentException: hostname can't be null
	at java.net.InetSocketAddress.checkHost(InetSocketAddress.java:149)
	at java.net.InetSocketAddress.<init>(InetSocketAddress.java:216)
	at redis.clients.jedis.Connection.connect(Connection.java:119)
	at redis.clients.jedis.BinaryClient.connect(BinaryClient.java:54)
	at redis.clients.jedis.BinaryJedis.connect(BinaryJedis.java:1657)
	at com.yahoo.ycsb.db.RedisClient.init(RedisClient.java:71)
	at com.yahoo.ycsb.DBWrapper.init(DBWrapper.java:77)
	at com.yahoo.ycsb.ClientThread.run(Client.java:289)
2016-02-09 21:28:02:792 0 sec: 0 operations; est completion in 0 seconds 
2016-02-09 21:28:02:795 0 sec: 0 operations; est completion in 106751991167300 days 15 hours 
[OVERALL], RunTime(ms), 21.0
[OVERALL], Throughput(ops/sec), 0.0

**Comment by kruthar** [comment]
Great thanks.

```
./bin/ycsb load redis -s -P workloads/workloada
```

This is wrong, hence the Exception.

```
./bin/ycsb load redis -s -P workloads/workloada -p "redis.host=127.0.0.1" -p "redis.port=6379"
```

This is correct, hence the successful output. You can tell it is successful when you look at the last part of the final output:

```
[INSERT], Operations, 1000.0
[INSERT], AverageLatency(us), 710.712
[INSERT], MinLatency(us), 431.0
[INSERT], MaxLatency(us), 10927.0
[INSERT], 95thPercentileLatency(us), 1506.0
[INSERT], 99thPercentileLatency(us), 1988.0
[INSERT], Return=OK, 1000
```

This is saying that 1000 insert operations and 1000 insert operations finished with the ```OK``` status.

**Comment by vishakha94** [comment]
How can I correct the error?
I dont understand why it (**./bin/ycsb load redis -s -P workloads/workloada**) says "host can't be null" when the previous command (**./bin/ycsb load redis -s -P workloads/workloada -p "redis.host=127.0.0.1" -p "redis.port=6379**) which I run , runs successfull

**Comment by kruthar** [comment]
These commands are not related to each other, the configuration you use for one command exists only for that command and does not affect future commands.

If you want to save configuration settings you can put all of your ```-p``` additions into a configuration file and reference it like you would the workload file like so:

```
./bin/ycsb load redis -s -P workloads/workloada -P path/to/myconfig
```

But remember you have to reference all configs on all commands, they are not stored in YCSB anywhere.

**Comment by vishakha94** [comment]
Thank you so much @kruthar

**Comment by kruthar** [comment]
NP, is this ok to close?

**Comment by vishakha94** [comment]
Sure.
